### PR TITLE
Sparse l1 descent

### PR DESCRIPTION
**Pull request by ftramer**
This PR adds a new type of Projected Gradient Descent attack tailored to the l1-norm, called the Sparse L1 Descent Attack. It deviates from previous PGD implementations in that it doesn't rely on the steepest descent direction, which is too sparse in the l1-case. Instead, it allows the caller to control the sparsity of the gradient updates. Experiments on MNIST and CIFAR10 show that with gradient updates that are ~99% sparse, we get a very efficient attack that rivals EAD in strength, yet is much more efficient.

The PR also generalizes PGD's random-start step to multiple norms. The current PGD implementation was using a random-start tailored to the l-inf norm even when ord=2.

A crucial optimization used in the Sparse L1 Descent Attack is to compute update steps by ignoring input components that would be clipped out anyway, thus focusing the attack budget on components where progress can still be made. This optimization isn't useful for the l-inf case, but can help in the l-2 case, so I added it to PGD and FGM for good measure.

### CODE REVIEW

File: cleverhans/attacks/momentum_iterative_method.py [hunk 1]
```diff
@@ -38,7 +38,7 @@ def __init__(self, model, sess=None, dtypestr='float32', **kwargs):
     self.feedable_kwargs = ('eps', 'eps_iter', 'y', 'y_target', 'clip_min',
                             'clip_max')
     self.structural_kwargs = [
-        'ord', 'nb_iter', 'decay_factor', 'sanity_checks']
+        'ord', 'nb_iter', 'decay_factor', 'sanity_checks', 'clip_grad']
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/momentum_iterative_method.py]: This attack inherits from PGD so it needs the same kwargs. I'm pretty sure some tests broke without this.

File: cleverhans/attacks/projected_gradient_descent.py [hunk 2]
```diff
@@ -175,7 +177,7 @@ def parse_params(self,
                    clip_max=None,
                    y_target=None,
                    rand_init=None,
-                   rand_minmax=0.3,
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/projected_gradient_descent.py]: This argument seems to have been added specifically for the l-inf case. We could make this a more general argument that specifies the size (in either linf or l2 norm) of the initial random start. But does anyone really use this? I don't see a reason why you wouldn't start from a random point within the full linf or l2 ball. Unless this can be used to implement R+FGSM?

File: cleverhans/attacks/projected_gradient_descent.py [hunk 3]
```diff
@@ -206,17 +211,14 @@ def parse_params(self,
     if rand_init is None:
       rand_init = self.default_rand_init
     self.rand_init = rand_init
-    if self.rand_init:
-      self.rand_minmax = eps
-    else:
-      self.rand_minmax = 0.
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/projected_gradient_descent.py]: same as above

File: cleverhans/utils_tf.py [hunk 4]
```diff
@@ -371,27 +371,144 @@ def clip_eta(eta, ord, eps):
...
+  if ord == np.inf:
+    r = tf.random_uniform(shape, -eps, eps, dtype=dtype, seed=seed)
+  else:
+
+    # For ord=1 and ord=2, we use the generic technique from
+    # (Calafiore et al. 1998) to sample uniformly from a norm ball.
+    # Paper link (Calafiore et al. 1998):
+    # https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=758215&tag=1
+    # We first sample from the surface of the norm ball, and then scale by
+    # a factor `w^d` where `w~U[0,1]` is a standard uniform random variable and
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/utils_tf.py]: thanks, that's a typo

File: cleverhans/utils_tf.py [hunk 5]
```diff
@@ -371,27 +371,144 @@ def clip_eta(eta, ord, eps):
...
+    # `d` is the dimension of the ball. In high dimensions, this is roughly
+    # equivalent to sampling from the surface of the ball.
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/utils_tf.py]: I think this is correct. I.e., we first sample a vector that lies on the surface of the ball (i.e., it has norm exactly epsilon) and then we scale it to get a uniform vector within the ball.

File: cleverhans/utils_tf.py [hunk 6]
```diff
@@ -371,27 +371,144 @@ def clip_eta(eta, ord, eps):
...
+
+      x = random_laplace((shape[0], dim), loc=1.0, scale=1.0, dtype=dtype,
+                         seed=seed)
+      norm = tf.reduce_sum(tf.abs(x), axis=-1, keepdims=True)
+
+      w = tf.pow(tf.random.uniform((shape[0], 1), dtype=dtype, seed=seed),
+                 1.0 / tf.cast(dim, dtype))
+      r = eps * tf.reshape(w * x / norm, shape)
+
+    else:
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/utils_tf.py]: sounds good

File: cleverhans/attacks/sparse_l1_descent.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+  abs_grad = tf.reshape(tf.abs(grad), (-1, dim))
+
+  # if q is a scalar, broadcast it to a vector of same length as the batch dim
+  q = tf.cast(tf.broadcast_to(q, tf.shape(x)[0:1]), tf.float32)
+  k = tf.cast(tf.floor(q / 100 * tf.cast(dim, tf.float32)), tf.int32)
+
+  # `tf.sort` is much faster than `tf.contrib.distributions.percentile`.
+  # For TF <= 1.12, use `tf.nn.top_k` as `tf.sort` is not implemented.
+  if 'sort' in dir(tf):
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/sparse_l1_descent.py]: sounds good

File: cleverhans/attacks/sparse_l1_descent.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+    if self.y is not None and self.y_target is not None:
+      raise ValueError("Must not set both y and y_target")
+
+    if self.clip_grad and (self.clip_min is None or self.clip_max is None):
+      raise ValueError("Must set clip_min and clip_max if clip_grad is set")
+
+    # The l1_grad_sparsity argument governs the sparsity of the gradient
+    # update. It indicates the percentile value above which gradient entries
+    # are retained. It can be specified as a scalar or as a 1-dimensional
+    # vector of the same size as the input's batch dimension.
```
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/sparse_l1_descent.py]: We could check that it's a vector using `.shape()` but we can't check that it has the right size as the batch size is undefined here.

File: cleverhans/attacks/momentum_iterative_method.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attacks/momentum_iterative_method.py]: sgtm

File: cleverhans/attacks/projected_gradient_descent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attacks/projected_gradient_descent.py]: Yeah I remember a paper using a ball of a smaller radius to do the random start. I'd rather keep it if you don't need to remove it and it's not too much work; since this version of the attack will be used to benchmark new implementations in tf2 / pytorch / jax as we move forward to implementations in new frameworks.

File: cleverhans/attacks/sparse_l1_descent.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (npapernot) [reply cleverhans/attacks/sparse_l1_descent.py]: no worries

**Comment by npapernot** [comment]
LGTM once comments are addressed

File: cleverhans/attacks/projected_gradient_descent.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ftramer) [reply cleverhans/attacks/projected_gradient_descent.py]: I've added this parameter back, renamed as rand_init_eps which should be more evocative of the general case. By default, this parameter takes the same value as eps.

**Comment by ftramer** [comment]
The latest commit should address all the raised comments.
